doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2310 @@
+We will make reference to this dialect throughout the document.
+
+
+## The Rust Toolchain {#the-rust-toolchain}
+
+A complete Rust toolchain consists of a few major parts:
+
+*   `rustc`, the Rust compiler[^3].
+*   `rustup`, the Rust toolchain installer[^4].
+*   `cargo`, a build system for Rust (though rustc can be invoked directly or by other tools).
+*   `std` and `core`, the standard libraries.
+
+The Rust toolchain is on a six-week release cycle, similar to Chrome's: every six weeks, a release branch is cut as the next beta, and after six weeks it becomes the next stable.
+Nightly Rust is cut from master every day; it is on nightly that unstable features can be enabled.
+Some unstable features[^5] are very useful for embedded, so it is not uncommon for embedded Rust projects to use a nightly compiler.
+
+`rustup` is used for managing Rust installations.
+This is mostly necessary because Rust releases too frequently for operating system package managers to keep up, and projects can pin specific versions of Rust.
+When the Rust toolchain is installed through rustup, components such as rustc and cargo will be aware of it; `rustc +nightly-2020-03-22` will defer to rustup to download and execute the `rustc` nightly built on March 22.
+A file named `rust-toolchain` in a project directory can be used to the same effect.[^6]
+
+Cargo is a build system/package manager for Rust.
+It can build projects (that is, directories with a `Cargo.toml` file in them) and their dependencies automatically.
+Individual units of compilation in Rust are called "crates", which are either static libraries[^7] (i.e., comparable to a `.a` file) or fully linked native binaries.
+This is in contrast to C, where each `.c` file generates a separate object file.[^8] Rust also does not have headers, though it provides a module system for intra-crate code organization that will be discussed later on.
+Tock boards are a good example of what a more-complex cargo file looks like: [https://github.com/tock/tock/blob/master/boards/opentitan/Cargo.toml](https://github.com/tock/tock/blob/master/boards/opentitan/Cargo.toml)
+
+Some useful `cargo` subcommands include:
+*   `cargo check` runs rustc's checks, but stops before it starts emitting code and optimizing it.
+    This is useful for checking for errors during development.
+*   `cargo build` builds a library or binary, depending on the crate type.
+*   `cargo clippy` runs the Rust linter, Clippy.
+*   `cargo doc --open` builds crate documentation and then opens it in the browser.
+*   `cargo fmt` runs the Rust formatter[^9].
+
+Also, the contents of the `RUSTFLAGS` environment variable are passed to `rustc`, as a mechanism for injecting flags.
+
+The Rust standard library, like libc, is smaller in embedded environments.
+The standard library consists of three crates: `core`[^10], `alloc`, and `std`.
```

Comment:
This doesn't visualize correctly in the github markdown. The `[` and `]` disappear.

Reply:
This Markdown file isn't intended for GH; it's going to show up on the OpenTitan docs website. You can preview it with `util/build_docs.py`.